packages/delegation-process/src/services/delegationService.ts

```diff
@@ -67,7 +104,291 @@ export const retrieveEserviceById = async (
 };
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-export function delegationServiceBuilder(readModelService: ReadModelService) {
+export function delegationServiceBuilder(
+  readModelService: ReadModelService,
+  dbInstance: DB,
+  pdfGenerator: PDFGenerator,
+  fileManager: FileManager
+) {
+  const repository = eventRepository(dbInstance, delegationEventToBinaryDataV2);
+
+  async function createDelegation(
+    {
+      delegateId,
+      eserviceId,
+      kind,
+    }: {
+      delegateId: TenantId;
+      eserviceId: EServiceId;
+      kind: DelegationKind;
+    },
+    { authData, logger, correlationId }: WithLogger<AppContext>
+  ): Promise<Delegation> {
+    const delegatorId = unsafeBrandId<TenantId>(authData.organizationId);
+
+    logger.info(
+      `Creating a delegation for tenant ${delegateId} by ${
+        kind === delegationKind.delegatedConsumer ? "consumer" : "producer"
+      } ${delegatorId}`
+    );
+
+    assertDelegatorIsNotDelegate(delegatorId, delegateId);
+
+    const delegator = await retrieveTenantById(readModelService, delegatorId);
+    const delegate = await retrieveTenantById(readModelService, delegateId);
+
+    assertTenantAllowedToReceiveDelegation(delegate, kind);
+    await assertDelegatorAndDelegateIPA(delegator, delegate);
+
+    await retrieveEserviceById(readModelService, eserviceId);
+
+    if (kind === delegationKind.delegatedProducer) {
+      const eservice = await retrieveEserviceById(readModelService, eserviceId);
+      assertDelegatorIsProducer(delegatorId, eservice);
+    }
```

Comment:
Even better, we could just retrieve it inside the condition since it's not needed elsewhere!

Reply:
The check for eservice existence is needed in both cases! Otherwise, it could be possible to create a delegation for a non-existing eservice - I would retrieve it outside the if so that we get the eservice not found error

Reply:
You're actually right, done [here](https://github.com/pagopa/interop-be-monorepo/pull/1259/commits/72bcc9e90f677bc1dbe06a2058bfe2fc243cefbd)